### PR TITLE
WIP: Automated batch creation

### PR DESCRIPTION
**Pull request by jkilpatr**
Prototype for automated batch creation.

Efficiency remains a question, as does the impact on the tests.

### CODE REVIEW

File: module/x/gravity/abci.go [hunk 1]
```diff
@@ -39,6 +40,22 @@ func createValsets(ctx sdk.Context, k keeper.Keeper) {
...
 }
 
+func createBatches(ctx sdk.Context, k keeper.Keeper) {
+	// Auto BatchRequest Creation.
+	// This function automatically requests batches of all token types every 5 blocks
+	// thanks to the more profitable batches rule (see batch creation spec) new batches
+	// will not be created unless the sum of their fees is greater than the previous batch
+	// that has not yet been executed. Eventually batches will either time out, or a
+	// batch profitable enough to submit will be generated.
+	if ctx.BlockHeight()%5 == 0 {
```
Comment (jtremback) [inline module/x/gravity/abci.go:50]: This should probably be a param

File: module/x/gravity/abci.go [hunk 2]
```diff
@@ -39,6 +40,22 @@ func createValsets(ctx sdk.Context, k keeper.Keeper) {
...
+		tokens := k.GetEthereumOriginatedErc20Tokens(ctx)
```
Comment (jtremback) [inline module/x/gravity/abci.go:51]: Why only ethereum originated?

File: module/x/gravity/keeper/ethereum-originated.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
+package keeper
```
Comment (jtremback) [inline module/x/gravity/keeper/ethereum-originated.go:1]: What's this all about?

**Review by jtremback — COMMENTED** [review]
I don't understand what the ethereum originated stuff has to do with anything. All tokens are batched the same in Gravity

File: module/x/gravity/abci.go (hunk 2 above)
Comment (jkilpatr) [reply module/x/gravity/abci.go:51]: Cosmos originated assets that have been bridged are a separate list, figured I would test this first but then ran out of time.

File: module/x/gravity/keeper/ethereum-originated.go (hunk 3 above)
Comment (jkilpatr) [reply module/x/gravity/keeper/ethereum-originated.go:1]: didn't want to make keeper.go any longer decied to try expanding it like we did for batch.go, not sure I like the pattern.

**Comment by jkilpatr** [comment]
> I don't understand what the ethereum originated stuff has to do with anything. All tokens are batched the same in Gravity

The problem is how do we know what Ethereum assets have been bridged? do we loop through all the denoms already on the chain? Previously we implicitly got the bridged asset name when the user requested it and we only had to check that it was in fact correct. But now we need to know each asset bridge from cosmos to ethereum (easy we have a list already) and each asset brought over from Ethereum (harder, we don't have a list except implicitly in the bank module)

Overall I'm not happy with the level of computation that seems to be coming into this implementation. Previously we got to offload this extra index (minor) and the extra computation of attempting to create a batch (lots of reads and potential writes) to be paid for by a relayer. The realyer may do a million computations determining that it's not worth it to request a batch and that's free, the chain doesn't have to pay for it. But now every validator must perform the same calculation pretty often.

**Comment by jkilpatr** [comment]
closed for now.

**Pull request closed without merging**